### PR TITLE
Point volume projection fixes

### PR DESCRIPTION
**Pull request by anilyil**
## Purpose
The projection routine used to embed points in volumes had 2 shortcomings:
1. When the linesearch is triggered, it would predict an unreasonably small step size. Now I just backtrack.
2. The convergence check was relying on the magnitude of the parametric derivative, which can be small based on control point locations. This was causing issues with the pygeo embedding where pyspline would think the projection is fine but pygeo would reject the point since it was slightly above the tolerance required.

The changes has the potential to slightly change all cases that use FFDs. How this would show up is that due to the line search change, the code might take a slightly different convergence path, or due to the tolerance change, the code might do an extra iteration. The changes should be on the order 1e-10 or less, but it is still important to note. Within a case, it does not affect derivative accuracy, but comparing against old cases, it will affect the recorded functions (e.g. our regression tests).

Expected time until merged:
About a week or two.

## Type of change

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Checklist

- [ ] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [ ] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Comment by anilyil** [comment]
Do any of you have a simple test we can include for the embedding test @marcomangano @sseraj ? Just a small FFD will be fine, we can manually add a handful of coordinates that would have been problematic with the original implementation

**Comment by marcomangano** [comment]
@anilyil I don't have a small example but I am pretty sure I could try and reproduce the previous issue I fixed at the pyGeo level in [this PR](https://github.com/mdolab/pygeo/pull/116).
There are a few changes in the code that it is probably quicker to discuss offline, do you and @sseraj have time later today or early next week? It looks we'll also need to retrain the tests for both pySpline and pyGeo. This could also be a good opportunity to bring up again regression tests on a full optimization problem, including checking the final result and not just the initialization as we are doing now.

**Review by sseraj — COMMENTED** [review]
@marcomangano Let's discuss this next week. I summarized my thoughts below:

I think changing the exit criterion makes sense. We should also make a PR on pyGeo to remove `eps` and only have `embTol`. I checked that this change alone does not break the regression tests.

I'm not convinced about the line search change. Does the quadratic approximation always lead to small steps? I think we should backtrack only if the quadratic approximation is poor. Ideally, this should not break the regression tests.

**Comment by anilyil** [comment]
I am fine with adding a catch that changes it to backtracking if quadratic approx fails (like step size less than 0.1)? I am surprised changing the exit condition is not what broke the tests. I was expecting the linesearch to be inactive for most of the cases we are doing. Regardless of the tests getting broken or not, real world applications will change if we change the tolerance check.

**Comment by anilyil** [comment]
This turned out to be a much bigger headache than I thought... The original solver had a bug that is baked into all of our tests. See this termination criteria here: https://github.com/mdolab/pyspline/blob/7f0ff6d993eb7b5c0299cc169e70f6d360aaa68a/src/projections.F90#L565-L569 

The code checks if the current iteration has differed by eps from the previous iteration. If the detected change is less than eps, the code keeps the *previous* point and exits. This was originally done to detect points that are outside the volume and avoid running all the way until max iterations are reached. 

However, there are several cases where a situation like this happens: Say the eps value is set to 1e-10. The current point is around 1.1e-10. The next iteration the solver takes puts the point right inside eps; however, because the previous point was already very close, this update is very small as well, more importantly, below eps. So the code thinks this point got stuck and is not improving, and keeps the original point at 1.1e-10 convergence. The correct solution is to take the new point since it satisfies the tolerance, whereas the old one does not. The new algorithm avoids this check. I have actually revised the exit criteria quite a bit; I am also aiming to detect points that are outside the volume and terminate if we are not making progress towards the point by checking the projected gradient. 

The downside of this fix is that we now have 2 tests that fail in ADflow. The mismatch is tiny; the test `test_adjoint.py:TestAdjoint_3_rans_tut_wing.test_adjoint` results in:
```
Mismatched elements: 1 / 72 (1.39%)
Max absolute difference: 2.48547281e-08
Max relative difference: 1.37676067e-10
```
and the test `test_adjoint.py:TestAdjoint_1_euler_scalar_JST_tut_wing.test_adjoint` results in:
```
Mismatched elements: 2 / 72 (2.78%)
Max absolute difference: 1.72756245e-08
Max relative difference: 3.53686769e-10
```
These mismatches might be due to me running this on an ARM machine, but the tests do pass when I have pyspline/main compiled. The solution to this will be re-training these two adflow tests. 

Another big change that was needed here was the volume test in pyspline. Previously, it was using a least squares fit based volume using data points. The data points when plotted in 3D looks like this:
actual_data

However, the fit volume results in this hot mess:
failed_fit

Our projection and evaluation tests were trained with this, which was essentially just making sure the code fails exactly the same way and any change would cause them to fail. 

To address this issue, I have modified the test to use the "data" array as control points while creating the b-spline volume. I still loop over a range of b-spline orders in 3 directions, but the volume itself is created directly by passing control points. Then the test evaluates a range of points inside the volume and runs the projection tests with these. I also test points that are outside the domain to check if the solver can still find the nearest point in the domain.

Now that I modified the test, some methods related to getting information from the underlying data used to fit a volume is not tested. I have created an issue #49 about the volume interpolation and ideally, we need to add a new test that actually does interpolate correctly to address this. I don't have time to fix this now. I do not know if anyone is using this feature. 

Finally, I am adding a challenging FFD test to pyGeo in https://github.com/mdolab/pygeo/pull/157. The FFD is based on @sseraj's cases. The points added to the FFD does not embed properly without the solver fixes in this PR, and they do embed correctly with the fixes here.

**Comment by sseraj** [comment]
To dos:
- [ ] Look at `val_diff` check
- [ ] Retrain select ADflow tests

**Review by marcomangano — COMMENTED** [review]
Just adding some comments to keep track of the offline discussion. Very nice work.

Some additional things we could include in this PR:
- For `test_volumes.py`
    - update the docstrings of `run_volume_test`
    - update to f-strings

File: src/projections.F90 [hunk 1]
```diff
@@ -488,7 +490,25 @@ subroutine point_volume(x0, tu, tv, tw, ku, kv, kw, coef, nctlu, nctlv, nctlw, n
...
+        ! and the default projection tolerance we use in pygeo is 1e-10.
+        ! If the FFD derivative is on the order of 1e-2 or less,
+        ! the newton will quit when eps is about 1e-12, but R can still be above 1e-10.
+        ! The newton solver quits successfully in this case,
+        ! however, the higher level point embedding routine thinks
+        ! we are not inside the b-spline volume since we could not reach the
+        ! 1e-10 tolerance. instead, we just check if we are within the
+        ! specified tolerance in physical coordinates.
+        ! if the point is physically outside the b-spline volume, the line search fail
+        ! check below will catch that case and terminate early.
```
Comment (sseraj) [inline src/projections.F90:510]: We don't have the line search fail check anymore right? Can you rephrase this part?

File: tests/reg_tests/test_volumes.py [hunk 2]
```diff
@@ -2177,18 +2158,53 @@ def regression_test(self, handler, solve=False):
...
+                    test_name = f"points inside volume with ku={ku}, kv={kv}, kw={kw}"
+                    run_project_test(volume, handler, test_name, pts=test_pts)
+
+                    # also test a few points outside
+                    test_name = f"points outside volume with ku={ku}, kv={kv}, kw={kw}"
+                    test_pts_outside = np.array(
+                        [
+                            [1.0, 1.0, 1.0],
+                            [0.15, 0.017, 0.39],
+                            [-0.120510712442222084, 0.0551918480620798083, -0.1],
```
Comment (sseraj) [inline tests/reg_tests/test_volumes.py:2207]: Can you add some comments on which case each of these points tests?

File: tests/reg_tests/test_volumes.py [hunk 3]
```diff
@@ -22,39 +22,20 @@ def run_volume_test(volume, handler, test_name):
...
         handler.root_add_val("{} evaluate point {}".format(test_name, pt), volume(pt[0], pt[1], pt[2]), tol=1e-10)
 
-    # Get value corner
-    for i in range(8):
-        handler.root_add_val("{} getOrigValueCorner({})".format(test_name, i), volume.getOrigValueCorner(i))
-
     # Get Value edge
     for i in range(12):
         handler.root_add_val("{} getValueEdge({}, 0.25)".format(test_name, i), volume.getValueEdge(i, 0.25))
         handler.root_add_val("{} getValueEdge({}, 0.75)".format(test_name, i), volume.getValueEdge(i, 0.75))
```
[the remark this answers is not in the document]
Comment (anilyil) [reply tests/reg_tests/test_volumes.py:29]: Yes. When you initialize the volume with control points directly, these functions do not work.

File: tests/reg_tests/test_volumes.py [hunk 4]
```diff
@@ -2177,18 +2158,53 @@ def regression_test(self, handler, solve=False):
...
             for kv in [3, 4]:
                 for kw in [2, 4]:
-                    for nCtlu in [4, 6]:
-                        for nCtlv in [8]:
-                            for nCtlw in [5, 7]:
-                                test_name = "volume with ku={}, kv={}, kw={}, nCtlu={}, nCtlv={}, nCtlw={}".format(
-                                    ku, kv, kw, nCtlu, nCtlv, nCtlw
-                                )
-                                volume = Volume(X=X, ku=ku, kv=kv, kw=kw, nCtlu=nCtlu, nCtlv=nCtlv, nCtlw=nCtlw)
-                                volume.recompute()
```
[the remark this answers is not in the document]
Comment (anilyil) [reply tests/reg_tests/test_volumes.py:2191]: Not sure, maybe to have everything up to date. The recompute call does have a few intermediate calculations that might be used later on.

File: tests/reg_tests/test_volumes.py [hunk 5]
```diff
@@ -2177,18 +2158,53 @@ def regression_test(self, handler, solve=False):
...
-                                run_volume_test(volume, handler, test_name)
-                                run_project_test(volume, handler, test_name)
-                                io_test(volume, handler)
+
```
[the remark this answers is not in the document]
Comment (anilyil) [reply tests/reg_tests/test_volumes.py:2176]: Yes, we are using the same set of control points. The knot vector is adjusted for different spline orders, but effectively, we always use the full set of control points. To me this is good enough. Testing fewer control points would not really change the behavior in any meaningful way.

File: src/projections.F90 [hunk 6]
```diff
@@ -500,9 +520,17 @@ subroutine point_volume(x0, tu, tv, tw, ku, kv, kw, coef, nctlu, nctlv, nctlw, n
 
         !Check that this is a descent direction -
         !otherwise use the negative gradient
-        if (pgrad >= 0.0) then
-            update = -grad / NORM2(grad)
+        if (pgrad .gt. 0.0) then
+            update = -grad / (NORM2(grad) + eps)
```
[the remark this answers is not in the document]
Comment (anilyil) [reply src/projections.F90:524]: Not just stability, w/o that eps check, the step can be zero and that division results in a NaN. I am thinking the exit criteria based on the conditions we discussed today will render that unnecessary, but I would still like to have it since this is one location a NaN might show up in the computation. If the step gets screwed up large due to division by zero, the linesearch just backtracks until its tiny anyways.

File: src/projections.F90 [hunk 7]
```diff
@@ -544,32 +572,35 @@ subroutine point_volume(x0, tu, tv, tw, ku, kv, kw, coef, nctlu, nctlv, nctlw, n
...
             else
-                ! c = nfval - fval - pgrad * step is always positive since
-                ! nfval - fval > pgrad * wolfe * step > pgrad * step
                 c = ((nfval - fval) - pgrad * step)
-                step = -step * step * pgrad / (2.0 * c)
+                quad_step = -step * step * pgrad / (2.0 * c)
                 ! This update is always less than the original step length
+                ! however, it might be too small... in that case,
+                ! just backtrack since the quad. approx is probably not too good.
+                if ((quad_step .lt. eps) .or. (quad_step .ne. quad_step)) then
```
[the remark this answers is not in the document]
Comment (anilyil) [reply src/projections.F90:582]: Yes, the quad step can get a NaN. I am not sure if a NaN triggers that first check as true. Just as a safety check.

File: src/projections.F90 [hunk 8]
```diff
@@ -544,32 +572,35 @@ subroutine point_volume(x0, tu, tv, tw, ku, kv, kw, coef, nctlu, nctlv, nctlw, n
...
-            end if
-        else
-            NLSFail = 0
-            ! Check if there has been no change in the coordinates
-            p_diff = NORM2(pt - newpt)
-            if (p_diff < eps) then
-                exit Iteration_loop
-            end if
-        end if
+        ! update the point
```
[the remark this answers is not in the document]
Comment (anilyil) [reply src/projections.F90:591]: It would cause difficult points to terminate early for no reason. The line search "failure" was based on it reaching the max iterations. @sseraj has several points that was quitting prematurely because of this check.

File: src/projections.F90 [hunk 9]
```diff
@@ -544,32 +572,35 @@ subroutine point_volume(x0, tu, tv, tw, ku, kv, kw, coef, nctlu, nctlv, nctlw, n
...
         pt = newpt
 
+        ! Check if there has been no change in the coordinates; we want to try to catch cases
+        ! where the point we want to project is outside the volume here
+        ! without running through all of the iterations.
+        ! we can test this by checking if the point stopped changing despite taking steps that
+        ! are larger than epsilon.
+        val_diff = NORM2(val - old_val)
+        if ((val_diff .lt. eps) .and. (step .gt. eps)) then
```
[the remark this answers is not in the document]
Comment (anilyil) [reply src/projections.F90:600]: I will fix this criteria by explicitly checking if we are at a bound and detect if we cannot make any more progress. Then I will remove this check alltogether.

File: src/projections.F90 (hunk 1 above)
Comment (anilyil) [reply src/projections.F90:510]: I rephrased it. It was actually unnecessarily long so I simplified it.

File: src/projections.F90 (hunk 9 above)
[the remark this answers is not in the document]
Comment (anilyil) [reply src/projections.F90:600]: This check is now removed.

File: tests/reg_tests/test_volumes.py (hunk 2 above)
Comment (anilyil) [reply tests/reg_tests/test_volumes.py:2207]: I changed the points here to explicitly test the face, edge, and corner cases.

**Comment by anilyil** [comment]
This is ready for the second round. The bounds check works as expected for the tests I have tried so far.

**Review by sseraj — APPROVED** [review]
Nice work
